rust_src/src/marker.rs

```diff
@@ -341,12 +337,121 @@ pub extern "C" fn marker_byte_position(marker: Lisp_Object) -> ptrdiff_t {
 /// whether BUFFER is a buffer object and return buffer pointer
 /// corresponding to BUFFER if BUFFER is live, or NULL otherwise.
 #[no_mangle]
-pub extern "C" fn live_buffer(buffer: Lisp_Object) -> *mut Lisp_Buffer {
-    let mut b = LispObject::from_raw(buffer).as_buffer_or_current_buffer();
+pub extern "C" fn live_buffer(buffer: Lisp_Object) -> Option<LispBufferRef> {
+    let b = LispObject::from_raw(buffer).as_buffer_or_current_buffer();
     if b.is_live() {
-        b.as_mut()
+        Some(b)
+    } else {
+        None
+    }
+}
+
+/// Internal function to set MARKER in BUFFER at POSITION.  Non-zero
+/// RESTRICTED means limit the POSITION by the visible part of BUFFER.
+fn set_marker_internal(
+    marker: LispObject,
+    position: LispObject,
+    buffer: LispObject,
+    restricted: bool,
+) -> LispObject {
+    let buf = live_buffer(buffer.to_raw());
+    let mut m = marker.as_marker_or_error();
+
+    // Set MARKER to point nowhere if BUFFER is dead, or
+    // POSITION is nil or a marker points to nowhere.
+    if position.is_nil()
+        || position
+            .as_marker()
+            .map_or(false, |m| !m.buffer().is_some()) || !buf.is_some()
+    {
+        unsafe {
+            unchain_marker(m.as_mut());
+        }
+
+    // Optimize the special case where we are copying the position of
+    // an existing marker, and MARKER is already in the same buffer.
+    } else if position.as_marker().map_or(false, |p| p.buffer() == buf) && m.buffer() == buf {
```

Comment:
Consistency. This time around the marker is named `p` but it is named `m` in the previous if.

Reply:
I changed this to `p` because I thought it might be confusing with 2 `m`s and this way it's clear that it belongs to `position`. Even if you can see that the closure shadows `m`, I think it's easier to read.

Reply:
Fair enough.